server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MirroringMigrationService.java

```diff
@@ -0,0 +1,276 @@
+    private static final String PATH_LEGACY_CREDENTIALS_BACKUP = PATH_LEGACY_CREDENTIALS + ".bak";
+
+    private final ProjectManager projectManager;
+
+    @Nullable
+    private List<String> shortWords;
+
+    MirroringMigrationService(ProjectManager projectManager) {
+        this.projectManager = projectManager;
+    }
+
+    void migrate() throws Exception {
+        for (Project project : projectManager.list().values()) {
+            migrateCredentials(project);
+            // Update the credential IDs in the mirrors.json file.
+            migrateMirrors(project);
+        }
+        shortWords = null;
+    }
+
+    private void migrateMirrors(Project project) throws Exception {
+        final List<MirrorCredential> credentials = project.metaRepo().credentials().join();
+        final ArrayNode mirrors = getMetaData(project, PATH_LEGACY_MIRRORS);
+        if (mirrors == null) {
+            return;
+        }
+
+        final Set<String> mirrorIds = new HashSet<>();
+        for (JsonNode mirror : mirrors) {
+            if (!mirror.isObject()) {
+                logger.warn("A mirror config must be an object: " + mirror);
+                continue;
+            }
+            try {
+                migrateMirror(project, (ObjectNode) mirror, mirrorIds, credentials);
+            } catch (Exception e) {
+                logger.warn("Failed to migrate a mirror config: " + mirror, e);
+            }
+        }
+
+        // Back up the old mirrors.json file and don't use it anymore.
+        rename(project, PATH_LEGACY_MIRRORS, PATH_LEGACY_MIRRORS_BACKUP);
+    }
+
+    private void migrateMirror(Project project, ObjectNode mirror, Set<String> mirrorIds,
+                               List<MirrorCredential> credentials) throws Exception {
+        String id;
+        final JsonNode idNode = mirror.get("id");
+        if (idNode == null) {
+            // Fill the 'id' field with a random value if not exists.
+            id = generateRandomIdForMirror(project.name(), mirror);
+        } else {
+            id = idNode.asText();
+        }
+        id = uniquify(id, mirrorIds);
+        mirror.put("id", id);
+
+        fillCredentialId(mirror, credentials);
+        if (mirror.get("schedule") == null) {
+            mirror.put("schedule", DEFAULT_SCHEDULE);
+        }
+        mirrorIds.add(id);
+
+        final String jsonFile = mirrorFile(id);
+        project.metaRepo()
+               .commit(Revision.HEAD, System.currentTimeMillis(), Author.SYSTEM,
+                       "Migrate the mirror " + id + " in '" + PATH_LEGACY_MIRRORS + "' into '" + jsonFile +
+                       "'.",
+                       Change.ofJsonUpsert(jsonFile, mirror))
+               .get();
+    }
+
+    private static void fillCredentialId(ObjectNode mirror, List<MirrorCredential> credentials) {
+        final JsonNode credentialId = mirror.get("credentialId");
+        if (credentialId != null) {
```

Comment:
Central Dogma doesn't have transactions so some data is updated during migration, it could remain in `.bak`. I changed to call `rename()` right after reading metadata which users prevent updating the configuration from the JSON patch. 

But I have no idea how to prevent JSON upsert operation. We may need to stop replication just before rolling updates.

Reply:
> But I have no idea how to prevent JSON upsert operation. We may need to stop replication just before rolling updates.

Instead of the current approach of updating from a Leader directly, would it be possible to issue a custom command (i.e. `MigrateMirrorCommand`) ?

If we take a write lock while migrating, I guess there will be minimal issues regarding inconsistency.

Reply:
> We may need to stop replication just before rolling updates.

That's a good solution. 👍 

> would it be possible to issue a custom command (i.e. MigrateMirrorCommand)

That would also work. 👍

Reply:
> Instead of the current approach of updating from a Leader directly, would it be possible to issue a custom command (i.e. MigrateMirrorCommand) ?

It would be possible if we can 1) make the mirror ID an credential ID deterministic and 2) all replica has the same revision for all repositories.

One possible scenario is:

1. Stop writing for all replicas.
2. Wait until all pending logs are replayed.
3. Execute a special command `MigrateMirrorCommand` that should be written to the replication logs even if `writable` is false.
4. Trigger `MirroringMigrationService` and all replica produces identical results for migration.